### PR TITLE
Align titles #22376

### PR DESCRIPTION
**Pull request by AlextheGreat1509**
## PR Summary
We used the previous pull request: [[ENH]: add Figure.align_titles()](https://github.com/matplotlib/matplotlib/pull/22793) as inspiration. We fixed the issues associated with it and did some refactoring. This pull request adds support for aligning titles between subplots. All use cases work. We tried to add tests but were unable to make them show the same behaviour as expected even though our manual tests did. We used two different ways to compare and generate the images but still could not get the desired behaviour, even though the images were generated with the same function (savefig).

## PR Checklist

**Documentation and Tests**
- [WIP] Has pytest style unit tests (and `pytest` passes)
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] New plotting related features are documented with examples.

**Release Notes**
- [N/A] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [N/A] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

File: lib/matplotlib/tests/test_figure.py [hunk 1]
```diff
@@ -102,6 +103,38 @@ def test_align_labels_stray_axes():
...
+
+    axs[0][1].set_title('Title2', loc="left")
+    fig.align_titles()
+    fig.savefig("./result_images/test_figure/figure_align_titles")
+    compare_images("./lib/matplotlib/tests/baseline_images/figure_align_titles.png",
+                   "./result_images/test_figure/figure_align_titles.png", 0)
+
+
+## TODO add image comparison
+@image_comparison(['figure_align_titles_param'], extensions=['png', 'svg'],
```
Comment (ksunden) [inline lib/matplotlib/tests/test_figure.py:124]: The test failures that are directly related to this PR are due to the baseline images for this image comparison not being added to the repo

If it is passing locally, those two files (png, svg) should be added in the `baseline_images` folder

File: lib/matplotlib/tests/test_figure.py [hunk 2]
```diff
@@ -102,6 +103,38 @@ def test_align_labels_stray_axes():
...
+                                        "title": "Title"})
+    axs[0][0].imshow(plt.np.zeros((5, 3)))
+    axs[0][1].imshow(plt.np.zeros((3, 5)))
+    axs[1][0].imshow(plt.np.zeros((2, 1)))
+    axs[1][1].imshow(plt.np.zeros((1, 2)))
+
+    axs[0][1].set_title('Title2', loc="left")
+    fig.align_titles()
+    fig.savefig("./result_images/test_figure/figure_align_titles")
+    compare_images("./lib/matplotlib/tests/baseline_images/figure_align_titles.png",
```
Comment (ksunden) [inline lib/matplotlib/tests/test_figure.py:119]: I think we need to not use relative paths for this test

**Review by ksunden — COMMENTED** [review]
The other error (with stubtest) is because of the recently added type hints. Since you are adding a function, that also needs to be added to `figure.pyi`

For this it will look like:

```python
   def align_titles(self, axs: list[Axes] | None = ...) -> None: ...
```

And be placed right above `align_xlabels`, as it is here.

I'm in the process of writing up better developer docs for these, but they are new enough that we are still working out some details about how the iteration on that idea goes forth (This is in fact the first _new_ method I think since that got added)

The "nbagg" test failures are understood and unrelated to this PR

**Comment by SirSkillful** [comment]
First of all, thank you for looking into this! We really appreciate it!
Secondly, I wanted to show an example of the error we are having with the tests:
If you run the following code in Jupyter notebook or just a normal python script, the image generated by fig.savefig() and plt.savefig() will look like this:
**Code:**
fig, axs = plt.subplots(2, 2,
                        subplot_kw={"xlabel": "x", "ylabel": "",
                                    "title": "Title"})
axs[0][0].imshow(plt.np.zeros((5, 3)))
axs[0][1].imshow(plt.np.zeros((3, 5)))
axs[1][0].imshow(plt.np.zeros((2, 1)))
axs[1][1].imshow(plt.np.zeros((1, 2)))

axs[0][1].set_title('Title2', loc="left")
fig.draw_without_rendering()
fig.align_titles()
fig.draw_without_rendering()
fig.savefig("fig_align_titles.png")

**Generated image**:
fig_align_titles


The problem we are facing now is, that the image generated by pytest when fig.savefig() is used looks like this:
**Image generated by pytest**:
test_image

When align_labels() is used though, the labels in images generated by pytest do show aligned labels. We are not sure what else we should do.

**Comment by jklymak** [comment]
The tests use the "classic" style by default.   For copying on recent matplotlib, I think `plt.style.use('classic')` should look the same.

**Comment by SirSkillful** [comment]
Is the 'classic' style rendered differently then? Have there been known workarounds or solutions?

**Comment by jklymak** [comment]
The test images were created before the new default style, so many older tests use the old style.  https://matplotlib.org/stable/gallery/style_sheets/style_sheets_reference.html

**Comment by ksunden** [comment]
Since these are _new_ test images, they should probably set to the new style by passing `style="default"` to the `image_comparison` decorator:

See this for an example:

https://github.com/matplotlib/matplotlib/blob/87a2b9d96ec0b9c86bd339ab8e138b8883fa6daa/lib/matplotlib/tests/test_polar.py#L10

**Comment by SirSkillful** [comment]
@jklymak That makes sense.
@ksunden Thanks for pointing that out! I didn't know you could do that.
Is this an allowed approach? If that is the case, we would change the tests to use the 'default' style.

**Comment by ksunden** [comment]
For new tests it is allowed, even encouraged. (Though if we can avoid image comparisons all together, that is even moreso encouraged... haven't looked too closely here to know how good such tests would be, and not too offended by more image tests, just they are a higher maintenance burden than non-image tests.)

It would be a big lift to change _old_ tests, though, and that is why its the default still...

There have been talks about explicitly setting old style on all image comparisons where it is not explicit, then transitioning to newer style being default (and thus the case for all new tests) but that is a slow process.

https://matplotlib.org/stable/devel/testing.html#writing-an-image-comparison-test

Also actually the preferred key is "mpl20" which is just a stable alias for the current "default" style, such that _if_ we change defaults again, the tests using it will retain current behavior. (It is currently just mapped to "default", and the similar alias "mpl15" is mapped to "classic")

**Comment by SirSkillful** [comment]
Thanks the info, I really appreciate it! Then we will update the tests asap :)

**Comment by jklymak** [comment]
FWIW, I think it's probably bad idea to have examples that are clearly broken in some respect - in this case with overlapping labels.  Either manually make enough space with subplots_adjusts or some such, or use constrained_layout.  In fact this should almost definitely get a test with constrained_layout.

**Comment by melissawm** [comment]
@AlextheGreat1509 do you need help addressing the CI failures here?

**Comment by rcomer** [comment]
Closing in favour of #27952.

**Pull request closed without merging**